src/ome_zarr_models/v04/multiscales.py

```diff
@@ -277,7 +285,63 @@ class MultiscaleGroupAttrs(Base):
     omero: Omero | None = None
 
 
+def _check_datasets_exist(data: MultiscaleGroup) -> MultiscaleGroup:
+    """
+    Check that the datasets referenced in the `multiscales` metadata are actually contained in this group.
+    """
+    attrs = data.attributes
+    flattened = data.to_flat()
+
+    for multiscale in attrs.multiscales:
+        for dataset in multiscale.datasets:
+            dpath = "/" + dataset.path.lstrip("/")
+            if dpath in flattened:
+                if not isinstance(flattened[dpath], ArraySpec):
+                    msg = (
+                        f"The node at {dpath} should be an array, "
+                        f"found {type(flattened[dpath])} instead"
+                    )
+                    raise ValueError(msg)
+            else:
+                msg = (
+                    f"Dataset {dataset.path} was specified in multiscale metadata, but no "
+                    "array with that name was found in the hierarchy. "
+                    "All arrays referenced in multiscale metadata must be contained in the group."
+                )
+                raise ValueError(msg)
+    return data
+
+
+def _check_array_ndim(data: MultiscaleGroup) -> MultiscaleGroup:
+    """
+    Check that all the arrays referenced by the `multiscales` metadata have dimensionality 
+    consistent with the number of axes defined in the metadata.
+    """
+    multimeta = data.attributes.multiscales
+    flat_self = data.to_flat()
+
+    for multiscale in multimeta:
+        multiscale_ndim = len(multiscale.axes)
+        for dataset in multiscale.datasets:
+            arr: ArraySpec = flat_self["/" + dataset.path.lstrip("/")]
```

Comment:
Perhaps if there's a performance cost to looking up the array metadata (which there will be for remote data), then we should think about merging `_check_array_ndim` and `_check_datasets_exist` to reduce the number of accesses?

But for now, I think it's clearer and easier to keep these as seaparate functions.

Reply:
all of the IO is done before these functions are called. the input to this function is a pydantic model with no handle to the zarr group. basically we do IO in one place (`from_zarr`), and everything after that is working with in-memory data